Autocomplete: add progress indicator and cancellation token

artworkad opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

When fetching data from a remote datasource the autocomplete component should visually indicate that it is awaiting results. Likewise the component should be able to cancel expensive asynchronous work before the result is returned, when OnSearchAsync and SelectOption are called.

The developer can choose between a circular progress indicator displayed inside the input field or a linear progress indicator displayed under the input field. If the input field has an icon on the right, the loading indicator will be shifted  to the left in order not to cover the icon.  The input field shrinks accordingly. If a clear icon is present, it will be hidden while progress indicator is displayed. Progress indicator color can be customized. The circular indicator is suitable for outlined autocomplete and the linear one for the default variant.

![ezgif-3-728ec09517](https://user-images.githubusercontent.com/524580/178247365-f5d55b75-34ca-4f5f-8a42-2c973db08ea1.gif)

![ezgif-3-7bea2f4f01](https://user-images.githubusercontent.com/524580/178248464-533ee611-73ac-4452-9ffb-0b37a322973c.gif)

![ezgif-3-5051269845](https://user-images.githubusercontent.com/524580/178247563-8b993cd4-ed3c-4941-8862-72d86bd17539.gif)

Closes https://github.com/MudBlazor/MudBlazor/issues/4061 and https://github.com/MudBlazor/MudBlazor/pull/3096

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

- Added unit test cases to existing AutocompleteTests.cs
- Tested visually

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

mckaragoz commented:
Does it have only circular progress?

artworkad commented:
> Does it have only circular progress?

In the third gif you can see linear progress.

Yomodo commented:
> We could have all three options: linear, circular and inside popover and let the developer choose.

Yeah that would be awesome. In many cases I find a linear progress bar too "heavy" and
only use it when I know it's a long running task.
A circular progress bar is small and is less distracting than a progress bar.

artworkad commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor:
```diff
@@ -3,7 +3,7 @@
 
 <CascadingValue Name="Standalone" Value="false" IsFixed="true">
-    <div class="mud-select mud-autocomplete">
+    <div class="@("mud-select " + (ShowProgressIndicator && IsLoading ? "mud-autocomplete mud-autocomplete--with-progress" : "mud-autocomplete"))">
```
What's a CssBuilder?

artworkad commented:
@mckaragoz @Yomodo @Mr-Technician 

Changes:

- Use CssBuilder
- Added the ability to display a progress message inside the popover
- Added tests

Overall the autocomplete component has following new parameters:

- ShowProgressIndicator, whether to show the progress indicator
- ProgressIndicatorType, whether to show linear or circular progress indicator
- ShowProgressInPopover, whether to show progress message inside popover
- ProgressInPopverMessage, the actual message to show

By default no indication of progress will be shown.

![ezgif-2-8873d073d0](https://user-images.githubusercontent.com/524580/179547858-bd13aefd-6160-4a44-b320-1282c973cd22.gif)

Yomodo commented:
@artworkad I like it, thank you!

just-the-benno commented:
Hi @artworkad, and thanks for putting in the effort. 

Your work is going in a good direction. Based on what we have learned, I'd like to make some suggestions.

 Remove ``ProgressIndicatorType`` and the corresponding enum. Use a ``RenderFragment`` instead and default it to a circular progress bar. This way, everyone who is unhappy with the circular, or wants some color, etc., can do it by putting in their own template. Because otherwise, the next PR is something like: "Adding Color to ProgressIndication in MudAutoComplete`` and we have another parameter. 
 
 The same applies to ``ProgressInPopverMessage``. While the message is acceptable, offer a ``RenderFragment`` to make it customizable. 
 
 Besides, your PR currently needs examples in the docs. 
 
 Regarding the ``CancellationToken``, I'd suggest you put an example together to check if this is the intended way it should work.  If I'm wrong, feel free to correct me.  For instance, you cancel the token when an item is selected. However, if the search has been completed and a user selects an item, there is nothing to cancel anymore? Isn't the use case more that there is currently an active async search, and you want to cancel it in favor of a "newer" one? 
 
 On a side note, if you want to make videos of the screen, I can recommend https://www.screentogif.com/

Mr-Technician commented:
@artworkad Can you take a look at benno's comments above and make changes as needed?

artworkad commented:
> @artworkad Can you take a look at benno's comments above and make changes as needed?

@Mr-Technician will do as soon as I return from vacation🌴on 8. August.

artworkad commented:
@just-the-benno @Mr-Technician I made changes according to benno's comment.

just-the-benno commented:
Hi @artworkad. I'm sorry for the delay. I'll review the PR soon.

What I still don't understand is the usage of the cancellation token. Shouldn't the token be referenced within an instance of ``MudAutocomplete`` and called from there when a new search is requested, and the old is not yet finished?

Yomodo commented:
The CancellationTokenSrc never gets disposed.

artworkad commented:
@Yomodo thx! I made changes to dispose the CancellationTokenSrc. 
@just-the-benno exactly, that's the idea behind it.

just-the-benno commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -851,5 +852,119 @@ public async Task Autocomplete_Should_ChangeAdornmentIcon()
+
+            comp.SetParam(p => p.ShowProgressIndicator, false);
+            comp.SetParam(p => p.DebounceInterval, 0);
+            comp.SetParam(p => p.SearchFunc, new Func<string, Task<IEnumerable<string>>>(async s =>
```
To show the "power" of the ``ShowProgressIndicator`` property, you could even set the render fragment here

artworkad replied:
Done

just-the-benno commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -851,5 +852,119 @@ public async Task Autocomplete_Should_ChangeAdornmentIcon()
+            comp.SetParam(a => a.Text, "Bar");
+
+            cancelled.Should().BeTrue();
+        }
```
See the comment on the review

artworkad replied:
Done

just-the-benno commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor:
```diff
@@ -22,8 +22,27 @@
+                    }
+                    else
+                    {
+                        <div class="@CircularProgressClassname">
```
Is the div here needed? Could you just add the ``@CircularProgressClassname`` to the ``MudProgressCircular ``?

artworkad replied:
I cannot think of a solution that avoids the div. I have to position the MudProgressCircular absolutely over the input field. This is what the div is for.

just-the-benno commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -119,6 +130,33 @@ public bool Dense
+        [Category(CategoryTypes.FormComponent.Appearance)]
+        public Color ProgressIndicatorColor { get; set; } = Color.Default;
+
+        private bool IsLoading { get; set; }
```
Is there a reason you choose to implement a property instead of a field? If you put it as field please start with an ``_`` and lowercase letter

artworkad replied:
No real reason, just a habit :D I've changed those to field.

just-the-benno commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -119,6 +130,33 @@ public bool Dense
+        [Category(CategoryTypes.FormComponent.ListBehavior)]
+        public Func<string, CancellationToken, Task<IEnumerable<T>>> SearchFuncWithCancel { get; set; }
+
+        private CancellationTokenSource CancellationTokenSrc { get; set; }
```
Is there a reason you choose to implement a property instead of a field? If you put it as field please start with an ``_`` and lowercase letter

artworkad replied:
Changed to field.

just-the-benno commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -277,6 +329,7 @@ public MudAutocomplete()
         {
             Adornment = Adornment.End;
             IconSize = Size.Medium;
+            CancellationTokenSrc = new CancellationTokenSource();
```
I'd remove the initialize here and rely on the ``CancelToken()`` method. In this way, you ensure that there is only one place where a new instance is created.

just-the-benno commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -365,6 +418,12 @@ protected override async Task UpdateValuePropertyAsync(bool updateText)
 
+        private void CancelToken()
+        {
+            CancellationTokenSrc.Cancel();
```
If you use the ``?`` operator (``CancellationTokenSrc?.Cancel();``) things become easier here

artworkad replied:
Thx! Done.

just-the-benno commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -623,6 +708,8 @@ private Task CoerceValueToText()
         protected override void Dispose(bool disposing)
         {
             _timer?.Dispose();
+            CancellationTokenSrc.Cancel();
```
I'd wrap this into a condition ``if(CancellationTokenSrc != null)`` and also, since ``.Cancel()`` can throw exceptions (https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.cancel?view=net-6.0#system-threading-cancellationtokensource-cancel), you should warp it additionally into a try-catch-block

artworkad replied:
Done

just-the-benno reviewed (changes requested):
Thanks, @artworkad, for your additional work. Your work is on an excellent track to be merged soon. You can see the comments for more details on minor things.

The "major" area of improvement should be around the ``CancellationToken`` aspect, starting with an example that is close to a common use case so that we can also contribute to good practice around async cancellation of tasks. I think that will also reveal how you want to handle the ``TaskCanceledException`` in ``MudAutocomplete``.  

My suggestion is you use the ``AmericanStatesController`` and creates a new method there that has a delay in it or add a param to the ``search`` endpoint that represents the delay time.

After that, I guess you can also create a more realistic unit test for this great addition of yours to MudBlazor.

just-the-benno commented:
@JonBunator, can you have a look at the changes regarding CSS? Anything from your side?

artworkad commented on src/MudBlazor.Docs/Pages/Components/Autocomplete/AutocompletePage.razor:
```diff
@@ -91,6 +91,22 @@
+                <Description>
+                    <MudText>When fetching data from a remote datasource MudAutocomplete can visually indicate that it is awaiting results. Following settings are available:</MudText>
+                    <br />
+                    <MudText><CodeInline>ShowProgressIndicator</CodeInline>: show circuar progress indicator.</MudText>
```
Fixed

artworkad commented on src/MudBlazor.Docs/Pages/Components/Autocomplete/Examples/AutocompleteProgressExample.razor:
```diff
@@ -0,0 +1,61 @@
+        </MudAutocomplete>
+    </MudItem>
+</MudGrid>
+
```
Done

artworkad commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -851,5 +852,119 @@ public async Task Autocomplete_Should_ChangeAdornmentIcon()
+            }));
+            comp.SetParam(a => a.Text, "Foo");
+
+            markup.Should().Contain("mud-progress-circular");
```
Done

artworkad commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -851,5 +852,119 @@ public async Task Autocomplete_Should_ChangeAdornmentIcon()
+            }));
+            comp.SetParam(a => a.Text, "Foo");
+
+            markup.Should().Contain("Loading...");
```
Done

artworkad commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -851,5 +852,119 @@ public async Task Autocomplete_Should_ChangeAdornmentIcon()
+            }));
+            autocompletecomp.SetParam(a => a.Text, "Foo");
+
+            markup.Should().Contain("Loading...");
```
Done

Yomodo commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -420,8 +419,8 @@ protected override async Task UpdateValuePropertyAsync(bool updateText)
         {
-            CancellationTokenSrc.Cancel();
-            CancellationTokenSrc = new CancellationTokenSource();
+            _cancellationTokenSrc?.Cancel();
```
Might I suggest a try/catch here as this can throw an `ObjectDisposedException`

artworkad replied:
Thx! Done.

artworkad commented:
@just-the-benno thank you for the valuable feedback. I think I covered everything in the last commits. Let me know if there are more changes to be made.

just-the-benno reviewed (approved):
Hi @artworkad, and thanks for your quick response. Nothing is stopping us from merging it. I'm curious if @JonBunator has something to say about your CSS additions.  

That said, I'd like to fiddle around with your PR and see if I can create a good example for the use of the cancellation token and maybe also refractor your test case for it a bit.  Currently it "feels" like something is missing, but I can't pinpoint it, so hopefully, that will help. ;) 

Or I confused something, which is also possible, So don't worry, It is planned for the weekend. Feel encouraged to keep me accountable for that goal.

just-the-benno commented:
Hi @artworkad 

I've added the example. Feel free to improve it so that your work is presented in the best possible light :)

If you have a look at the example, you see that the progress indicator disappears after the first cancellation. Does that seem like a minor bug that you could fix? 

![cancel-requests](https://user-images.githubusercontent.com/51370361/187376287-aa81e304-ecf7-4fac-a80f-071d420a6262.gif)

artworkad commented:
@just-the-benno your example is great. I would not change anything. Regarding the bug: when a search gets cancelled due to another search the SearchFuncWithCancel method will exit and the finally block will be called. This sets _isLoading to false, while the other search is still awaiting results. Here:

```
CancelToken();

try
{
	_isLoading = true;

	if (ProgressIndicatorInPopoverTemplate != null)
	{
		IsOpen = true;
	}

	if (ShowProgressIndicator || ProgressIndicatorInPopoverTemplate != null)
	{
		StateHasChanged();
	}
	
	if (SearchFuncWithCancel != null)
	{
		searched_items = (await SearchFuncWithCancel(Text, _cancellationTokenSrc.Token)) ?? Array.Empty<T>();
	}
	else
	{
		searched_items = (await SearchFunc(Text)) ?? Array.Empty<T>();
	}
}
catch (Exception e)
{
	Console.WriteLine("The search function failed to return results: " + e.Message);
}
finally
{
	_isLoading = false;
}
```

Do you have an idea, how to fix that? My first idea was to use a _isLoading stack instead of a flag. When starting a search we put true onto the stack, when finishing or cancelling a search we pop from the stack. We have a progress going on when there are items in the stack. What do you think?

Yomodo commented:
We use this pattern a lot throughout our application:

```csharp
public async ValueTask CompareDscAadGroupsAsync()
   {
      try
      {
         var cancellationToken = await InitializeCompareAadGroupsAsync();
         
         ...
      }
      catch (Exception ex)
      {
         ...
      }
      finally
      {
         await FinalizeCompareAadGroupsAsync();
      }
   }
```


```csharp
public async ValueTask CancelCompareAadGroupsAsync()
   {
      await FinalizeCompareAadGroupsAsync(stateHasChanged: false);
      SetCompareAadGroupsCanceledState();
      StateHasChanged();
   }


   async ValueTask<CancellationToken> InitializeCompareAadGroupsAsync()
   {
      await FinalizeCompareAadGroupsAsync(stateHasChanged: false);
      SetCompareAadGroupsCanceledState(isCanceled: false);
      SetCompareAadGroupsBusyState();
      SetDscAadGroupsCompareResults();

      _aadGroupCompareCts = new();
      return _aadGroupCompareCts.Token;
   }


   void SetCompareAadGroupsCanceledState(bool isCanceled = true)
   {
      ViewModel.GroupCompareCanceled = isCanceled;
   }


   void SetCompareAadGroupsBusyState(bool isBusy = true)
   {
      ViewModel.GroupCompareBusy = isBusy;
   }


   async ValueTask FinalizeCompareAadGroupsAsync(bool stateHasChanged = true)
   {
      await ValueTask.CompletedTask;
      try
      {
         if (_aadGroupCompareCts is not null)
         {
            using (_aadGroupCompareCts)
               _aadGroupCompareCts.Cancel();

            _aadGroupCompareCts = null;
         }
      }
      catch (Exception ex)
      {
         _aadGroupCompareCts = null;
         _logger.LogError($"{nameof(FinalizeCompareAadGroupsAsync)}: {ex.Message}");
      }

      SetCompareAadGroupsBusyState(isBusy: false);

      if (stateHasChanged)
         StateHasChanged();
   }
```

just-the-benno commented:
> Do you have an idea, how to fix that? My first idea was to use a _isLoading stack instead of a flag. When starting a search we put true onto the stack, when finishing or cancelling a search we pop from the stack. We have a progress going on when there are items in the stack. What do you think?

Hi @artworkad. I'm sorry for the delay. It has been quite a busy week for me.  

I like your approach. It is simple yet effective. Looking forward to seeing it implemented ;)

artworkad commented:
@just-the-benno @Yomodo the bug is fixed. Instead of setting the loading flag manually, we now check the status of the search task:

```
private Task _currentSearchTask;
private bool IsLoading => _currentSearchTask != null && !_currentSearchTask.IsCompleted;
```
Search method:
```
try
{
    if (ProgressIndicatorInPopoverTemplate != null)
    {
        IsOpen = true;
    }

    var searchTask = SearchFuncWithCancel != null ?
        SearchFuncWithCancel(Text, _cancellationTokenSrc.Token) :
        SearchFunc(Text);

    _currentSearchTask = searchTask;

    StateHasChanged();

    searched_items = await searchTask ?? Array.Empty<T>();
}
catch (Exception e)
{
    Console.WriteLine("The search function failed to return results: " + e.Message);
}
```

just-the-benno reviewed (approved):
Hi, @artworkad that is a very creative and excellent way to solve the problem. I added a small commit to filter certain exceptions that occur when a task is canceled to be falsely written as an error to the console. 

So - finally - this PR can be merged.

just-the-benno commented:
@henon, ready for your review

henon commented:
@artworkad: Thanks for the contribution. Check out https://www.screentogif.com/ for recording short clips.
